### PR TITLE
Resolve Clang warnings

### PR DESCRIPTION
**Pull request by hahnjo**
Building with current Clang `main` produced a number of warnings.

### CODE REVIEW

File: core/base/inc/TVirtualQConnection.h [hunk 1]
```diff
@@ -45,16 +45,6 @@ class TVirtualQConnection : public TList {
...
-   // We pass arrays as Lont_t*. In the template instance we can deduce their
-   // size, too.
-   template <class T, class = typename std::enable_if<std::is_array<T>::value>::type>
-   void SetArg(const T* val)
-   {
-      constexpr size_t size = sizeof(val)/sizeof(val[0]);
-      static_assert(size > 0, "The array must have at least one element!");
-      SetArg((Longptr_t*)val, size);
-   }
-
```
Comment (hahnjo) [inline core/base/inc/TVirtualQConnection.h:57]: @Axel-Naumann could you check my reasoning in the commit message? Am I missing something?

File: io/io/src/TMemFile.cxx [hunk 2]
```diff
@@ -177,9 +177,6 @@ TMemFile::TMemFile(const char *path, char *buffer, Long64_t size, Option_t *opti
    EMode optmode = ParseOption(option);
 
    if (NeedsToWrite(optmode)) {
-      Int_t mode = O_RDWR | O_CREAT;
-      if (optmode == EMode::kRecreate) mode |= O_TRUNC;
-
```
Comment (hahnjo) [inline io/io/src/TMemFile.cxx:182]: @pcanal it appears this variable is unused since the initial commit 88add5a1efb48d9de5b185198439edef56568d86. I thought about passing `mode` to `TMemFile::SysOpen`, but considering that it's been this way since 2011 I decided not to.

File: core/base/inc/TVirtualQConnection.h (hunk 1 above)
Comment (pcanal) [reply core/base/inc/TVirtualQConnection.h:57]: It is likely the original author (@vgvassilev ?) meant:
```
   template <class T, class = typename std::enable_if<std::is_array<T>::value>::type>
   void SetArg(const T val)
```
Cheers,
Philippe.

Comment (Axel-Naumann) [reply core/base/inc/TVirtualQConnection.h:57]: But @vgvassilev also says "We pass arrays as Lont_t*".

This is commit 95a27809bc55752d85a602e705d531b459de25b1 btw.

Comment (pcanal) [reply core/base/inc/TVirtualQConnection.h:57]: Well .. I assumed the comment applied to the inner part ... but actually it hinted to the fact that in order to get the size of the array, we need to use a different syntax.

See: https://godbolt.org/z/xqo4ah61h   :   with `SetArg(const T val)` and `SetArg(arr1);` `T` is actually 'MyClass*`.

See: https://godbolt.org/z/4Yovca1Kr  :   with `SetArg(const T* val)` and `SetArg(&arr1);` then it works (and the code needs to be updated to de-reference the pointer)

See: https://godbolt.org/z/dj3K1MhbG  :   we can get `SetArg(arr1);`  to work with: `SetArg(const T (&val)[N])`

File: io/io/src/TMemFile.cxx (hunk 2 above)
Comment (pcanal) [reply io/io/src/TMemFile.cxx:182]: Indeed, for consistency, it need to be passed to `SysOpen` but also, at least for now, `TMemFile::SysOpen` ignores the argument.

File: core/base/inc/TVirtualQConnection.h (hunk 1 above)
Comment (Axel-Naumann) [reply core/base/inc/TVirtualQConnection.h:57]: Agreed, @hahnjo - this function has never been called. Tagging @bellenot so he's aware of this in case it ever bubbles up in signal/slot.

**Comment by hahnjo** [comment]
Eh no, @pcanal's PR just contained a typo :sweat_smile: 

ping @Axel-Naumann @pcanal

**Comment by pcanal** [comment]
@hahnjo Sorry about that :(

**Review by pcanal — APPROVED** [review]
For the core/io/tree parts

**Comment by hahnjo** [comment]
@phsft-bot build

**Comment by hahnjo** [comment]
ping @Axel-Naumann

File: core/cont/inc/TBits.h [hunk 3]
```diff
@@ -147,12 +147,12 @@ class TBits : public TObject {
 
 inline Bool_t operator&(const TBits::TReference& lhs, const TBits::TReference& rhs)
 {
-   return (Bool_t)lhs & rhs;
+   return (Bool_t)lhs && rhs;
```
Comment (Axel-Naumann) [inline core/cont/inc/TBits.h:150]: While you're at it, can you rephrase this to 
```suggestion
   return ((Bool_t)lhs) && rhs;
```
(which is identical to the expression above, if I'm not mistaken, but clarifies the precedence)?

File: hist/hist/src/TH1.cxx [hunk 4]
```diff
@@ -6423,24 +6422,12 @@ Bool_t TH1::FindNewAxisLimits(const TAxis* axis, const Double_t point, Double_t&
          return kFALSE;
       xmin = xmin - range;
       range *= 2;
-      binsize *= 2;
-      // // make sure that the merging will be correct
```
Comment (Axel-Naumann) [inline hist/hist/src/TH1.cxx:6427]: For the future reader: this is 007bbbaaa4d635f9074bba7aaf84475d467c1b88

File: math/mathmore/test/testRandomDist.cxx [hunk 5]
```diff
@@ -484,14 +484,11 @@ void testMultinomial( R & r,int ntot, TH1D & h1, TH1D & h2) {
...
    const int nbins = h1.GetNbinsX();
    std::vector<double> p(nbins);
-   double psum = 0;
    for (int i = 0; i < nbins; ++i) {
       double x1 = h1.GetBinLowEdge(i+1);
       double x2 = x1 + h1.GetBinWidth(i+1);
       p[i] = ROOT::Math::normal_cdf(x2) -  ROOT::Math::normal_cdf(x1);
-      psum += p[i];
    }
-   //std::cout << " psum  = " << psum << std::endl;
```
Comment (Axel-Naumann) [inline math/mathmore/test/testRandomDist.cxx:494]: @lmoneta should we not assert that `psum` has a expected, golden value?

File: tmva/tmva/src/PDEFoam.cxx [hunk 6]
```diff
@@ -589,14 +589,11 @@ void TMVA::PDEFoam::Varedu(Double_t ceSum[5], Int_t &kBest, Double_t &xBest, Dou
          Double_t xMin=0.0; Double_t xMax=0.0;
          // Double loop over all pairs jLo<jUp
          for(Int_t jLo=1; jLo<=fNBin; jLo++) {
-            Double_t aswIn=0;  Double_t asswIn=0;
+            Double_t asswIn=0;
             for(Int_t jUp=jLo; jUp<=fNBin;jUp++) {
-               aswIn  +=     ((TH1D *)(*fHistEdg)[kProj])->GetBinContent(jUp);
```
Comment (Axel-Naumann) [inline tmva/tmva/src/PDEFoam.cxx:594]: @lmoneta could you crosscheck that `aswIn` is indeed not intended to be used? I.e. that the oversight is setting it and not with using it, where at the moment some other, wrong variable is used instead?

File: tree/tree/src/TTree.cxx [hunk 7]
```diff
@@ -4512,7 +4511,7 @@ void TTree::DropBuffers(Int_t)
          }
          TBasket* basket = (TBasket*)branch->GetListOfBaskets()->UncheckedAt(j);
          if (basket) {
-            ndrop += basket->DropBuffers();
+            basket->DropBuffers();
```
Comment (Axel-Naumann) [inline tree/tree/src/TTree.cxx:4514]: @pcanal should `DropBuffers()` return `ndrop`?

**Review by Axel-Naumann — COMMENTED** [review]
We need some help from @lmoneta and @pcanal here. Other than whhere I commented this looks great. Thank you so much, @hahnjo and `clang`!

**Comment by hahnjo** [comment]
ping @lmoneta @pcanal

File: tmva/tmva/src/PDEFoam.cxx (hunk 6 above)
Comment (lmoneta) [reply tmva/tmva/src/PDEFoam.cxx:594]: The variables are not used, I would however just comment them and leave in the code, since it is maybe some information that could be useful for debugging, if we remove the lines 594, 598 and 599  we will lose this.

**Review by lmoneta — APPROVED** [review]
Thank you Jonas for fixing these warnings, 
I have just a small comment on the changes in tmva/tmva/src/PDEFoam.cxx
Apologies for the delay in doing this, I missed the notification!

File: tmva/tmva/src/PDEFoam.cxx (hunk 6 above)
Comment (hahnjo) [reply tmva/tmva/src/PDEFoam.cxx:594]: Done, even though I'm not really a fan of leaving random debugging code commented out...
